
\section{Introduction}

The release of the RefSeq v1.0 genome assembly of bread wheat cultivar Chinese
Spring~\citep{IWGSC2018} provided the first detailed map of the 16Gb genome, bringing wheat genomic
resources closer to the standard enjoyed in rice and maize. 
%
Already, international efforts are in the process of assembling more than ten elite wheat varieties
to a similar standard. 
%
Within five years we can expect hundreds of high quality wheat genomes to be available. 
%
The challenge that remains is to connect these new resources to the wealth of research carried out
in the past few decades, and make this information accessible to researchers and breeders who may
lack the required bioinformatics skill-set.
%
Existing tools such as cMap~\citep{fang2003cmap} or genome
browsers~\citep{Stein2013GBrowse,Buels2016JBrowse} allow the visualisation of genetic maps and genome
assemblies, respectively, but are separate tools. Hence one cannot view a genetic map alongside a
physical genome sequence, for example. 
%
Wheat-focused databases such as GrainGenes and T3~\citep{Blake2016T3} provide a wealth of legacy
data but can be difficult to interface with. 
%
Pretzel aims to provide a framework that combines curated data stored in a back-end with a
front-end graphical user interface for visualisation and analysis via a modern web browser. 
%
\begin{figure*}
\centering
  \includegraphics[width=\textwidth]{pretzel.png}
\caption{
  Pretzel alignment showing a 90k genetic map on the far left against the IWGSC RefSeq v1.0 chromosome 1B pseudomolecule in the centre, 
  with chromosome 1D on the far right. 
  Green lines indicate direct links between features, red lines indicate links via alias. 
  The two pseudomolecules have the scaffold structure displayed in blue, loaded inside their axes.
}
\label{fig:01}
\end{figure*}
%
%\begin{methods}
\section{Methods}

Pretzel is built on several JavaScript frameworks: the front-end combines Ember.js with D3.js
  for real-time visualisation; the back-end is in Loopback.js connecting to a MongoDB database. 
%
The user interacts with Pretzel via a web page served by the front-end. 
%
The data model has been designed to be as general as possible while still enabling the richness of
  data to be captured. 
%
  The highest level data structure is a \textit{dataset}, which contains one or more \textit{blocks}.
%
For example, a dataset may describe a genetic map, where the blocks are individual linkage groups. 
%
  Within each block are a set of \textit{features} which are defined as intervals (possibly of zero length,
  which would define a single position) within the block. 
%
For example, a feature could be a molecular marker in a linkage group, or the position of a gene in
  a physical chromosome. 
%
Using only these concepts, a wide range of information can be captured: a quantitative trait locus
  (QTL) is simply an interval within a linkage group; pseudomolecule structure can be described by a
  set of intervals denoting scaffold start/end positions. 
%
As there are often multiple alternative names for molecular markers, we created the concept of
  \textit{alias}, which indicates that features with different names should be considered as the same when
  drawing alignments. 
%
This general concept also allows the association of syntenic genes between genomes.

Data can be uploaded through the web interface in CSV or JSON format (the native data structure in
  JavaScript); or uploaded via command line in JSON format to the back-end directly. 
%
This facilitates upload of many datasets. 
%
Upon upload, data is visible by that user only, unless they explicitly make it available to other
  users. 
%
In addition, we have created a pipeline, Pretzel Input Generator
  {{https://github.com/plantinformatics/pretzel-input-generator}}), to generate Pretzel-ready data
  from standard genome assembly formats in an automated way. This pipeline has been used to make
  available a set of publicly available genomes so users can quickly set up a Pretzel server
  populated with real data.

One of the most common questions asked in wheat research is what genes underlie a QTL interval
  defined by a genetic mapping experiment. 
%
With Pretzel, users with no bioinformatics training can quickly project their genetic map interval
  onto physical sequence and extract the genes in this region for downstream analysis. 
%
An additional challenge when fine mapping an interval is potential mis-assembly in the region
  of interest. Figure~\ref{fig:01} shows a genetic map aligned to a physical chromosome, as well as
  two syntenic chromosomes (chromosomes 1B and 1D) aligned via their syntenic genes (shown in red). 
%
By comparing the co-linearity of genes between sub-genomes, potential assembly problems can be
  flagged by noting when abrupt changes in co-linearity occur at scaffold break points. 
%
In this way, advanced curation of pseudomolecule structure can be achieved on any number of genomes.

The technology stack Pretzel is built on is lightweight and a user with the necessary dependencies
can clone, build and run the Pretzel code within minutes with a few simple commands. 
%
If run on an internet-facing machine, external users can then connect to the server.
%
To make the installation process even easier, we provide Docker container images on Docker Hub
(\href{https://hub.docker.com/r/plantinformaticscollaboration/pretzel/}{\nolinkurl{https://hub.docker.com/r/plantinformaticscollaboration/pretzel/}}).
%\end{methods}

\section{Conclusion}

We have developed a web-based visualisation and analysis tool that solves many challenges posed today in wheat research. 
%
Its development is ongoing and we believe the volume of data that will be publicly available in the next few years will only increase its utility and relevance.

\section*{Acknowledgements}

Blah di blah.

\section*{Funding}

This work has been supported by the Grains Research and Development Corporation project DAV00127.\vspace*{-12pt}

\bibliographystyle{natbib}

\bibliography{references}
